### DISCUSSION TITLE
Allow for &Parent usage in belonging_to

### DISCUSSION THREAD

**Discussion by valkum** [diesel-rs/diesel#3138] [Ideas]
Currently the `BelongingToDsl` bounds are restrictive in such a way that you are not able to pass a slice to `belonging_to` which contains references.
_The examples here are based on the types from diesel_test or similar constructed_

## Motivation
Sometimes you already grouped two related tables to get a Vec of tuples of different structures. E.g. 

```rust
let posts = Posts::belonging_to(&users).grouped_by(&users);
let res: Vec<(User, Vec<Post>)> = users.zip(posts);
``` 

If you want to fetch other related rows, you currently need to Clone the User.
While normally cloning objects of the size of a row is no intensive task, it is avoidable for some purposes.

I would expect the following to work (_Hint: it does not_):

```rust
let posts = Posts::belonging_to(&users).grouped_by(&users);
let res: Vec<(User, Vec<Post>)> = users.zip(posts);
let user_refs: Vec<&User> = users.iter().map(|x| x.0).collect::<Vec<_>>();
let likes = Like::belonging_to(&user_refs);

``` 
Obviously it seems odd in this small example, but if you have a larger project where you want to have a composable way of fetching e.g. Likes this seems like a missing feature.

## Problem
The trait bounds of BelongingToDsl restrict the expected referenced usage.
* &User does not implement Identifiable
* BelongsTo<&Parent> is not implemented for Child

## Idea

Provide a blanket impl for &T where T is Identifiable and Copy
```rust
impl<T> Identifiable for &T
where
    T: Identifiable + Copy,
{
    type Id = T::Id;

    fn id(self) -> Self::Id {
        T::id(*self)
    }
}
```

Ideally there would be a blanket impl for BelongsTo as well.
The problem here is that this (⬇️) overlaps with the tuple impls for `BelongsTo`.

```rust
impl<Parent, T> BelongsTo<&Parent> for T
where
    T: BelongsTo<T>,
{
    type ForeignKey = T::ForeignKey;
    type ForeignKeyColumn = T::ForeignKeyColumn;

    fn foreign_key(&self) -> Option<&Self::ForeignKey> {
        T::foreign_key(&self.0)
    }

    fn foreign_key_column() -> Self::ForeignKeyColumn {
        T::foreign_key_column()
    }
}
```

It seems there were similar issues with the `Insertable` trait, which got a `InsertableOptionHelper`.
Sadly, this is not 1to1 applicable to solve the impl overlap issue here.

Maybe someone with a deeper understanding can come up with such a way.

Another solution would be to add a `BelongsTo<&Parent>` impl to the derive macro, but I guess having a blanket impl is preferred (if possible) to avoid compilation blowup?

## Nice To Have
A way to pass an iterator to `T::belonging_to`. As belonging_to for slices already uses an Iterator to get and collect the ids, the nicest way would be to allow to pass something that implements Iterator<Item = &Parent>. But adding a `BelongingToDsl` for such a kind does overlap with the current normal `BelongingToDsl` implementations.

**Comment by weiznich** [comment]
Thanks for writing this proposal down :+1:

First of all: I personally will likely not work on this in the near future, but I might accept well written PR's.

With that out of the way, let me comment on specific parts of the proposal:

First of all I would like to thank you for writing down such a well structured proposal. That's always a good starting point.

> Provide a blanket impl for &T where T is Identifiable and Copy

We already provide a `Identifiable` impl for reference types. In fact `#[derive(Identifiable)]` generates only an impl for references: https://github.com/diesel-rs/diesel/blob/50c2c85032857a65e4a8d68dc851fa7ab91d65b2/diesel_derives/src/identifiable.rs#L38

This means this proposed wild card impl is likely not necessary and will likely collide with the existing derived ones.

> The problem here is that this (⬇️) overlaps with the tuple impls for BelongsTo.

We could workaround this overlapping impl issue by manually generating an impl for concrete types here. That would mean that we generate two `BelongsTo` impls via `#[derive(Associations)]` instead of the existing one here: https://github.com/diesel-rs/diesel/blob/50c2c85032857a65e4a8d68dc851fa7ab91d65b2/diesel_derives/src/associations.rs#L68

Basically just generate a second `impl #impl_generics diesel::associations::BelongsTo<&'a #parent_struct>` there. As this uses concrete types this should not trigger a conflicting impl error.


> A way to pass an iterator to T::belonging_to. As belonging_to for slices already uses an Iterator to get and collect the ids, the nicest way would be to allow to pass something that implements Iterator<Item = &Parent>. But adding a BelongingToDsl for such a kind does overlap with the current normal BelongingToDsl implementations.

Using a `Iterator` there seems like a nice idea, but I cannot see how we could have another generic impl there. What could be done is adding a impl for `Box<dyn Iterator<Item = &'a Parent>` as this is a concrete type. That's not ideal, but at least something.

**Reply by valkum** [reply]
Thanks for your fast reply.

We already tried different ways to implement the ideas.

> We already provide a `Identifiable` impl for reference types. In fact `#[derive(Identifiable)]` generates only an impl for references: 

The problem is: The trait bounds for `BelongingToDsl` require `&'a Parent` to implement `Identifiable`.
If you would pass a slice of references to fulfill full the trait bounds `&'a &Parent` would need to implement Identifiable which is currently not the case.

> This means this proposed wild card impl is likely not necessary and will likely collide with the existing derived ones.

In a current prototype with the `Copy` bound for `T` it works. But I guess that will break once a struct that uses the derive also implements `Copy`. Maybe we can come up with a different solution for that problem. I personally would avoid adding a `&&ConcreteType` impl to the derive macro.

> We could workaround this overlapping impl issue by manually generating an impl for concrete types here. That would mean that we generate two `BelongsTo` impls via `#[derive(Associations)]` instead of the existing one here: 

That works. Ideally we wanted to avoid deriving an additional impl for each derived struct but if that is an acceptable solution we will use this way in a possible PR.

> Using a Iterator there seems like a nice idea, but I cannot see how we could have another generic impl there. What could be done is adding a impl for Box<dyn Iterator<Item = &'a Parent> as this is a concrete type. That's not ideal, but at least something.

Good idea. We will test this idea and include it or a similar one in a possible PR.

**Reply by weiznich** [reply]
> In a current prototype with the Copy bound for T it works. But I guess that will break once a struct that uses the derive also implements Copy. Maybe we can come up with a different solution for that problem. I personally would avoid adding a &&ConcreteType impl to the derive macro.

We could just add a wild card impl for that without depending on `Copy`:

```rust
impl<'a, T> Identifiable for &'a &'a T
where &'a T: Identifiable 
{
 // …
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aa1591ad3cbae8969dfe0b62b6afe924)

Note: If we choose to go that route we need to do that soon, so that this is included in the upcoming 2.0 release. Adding this impl later on would be a breaking change.

Other than that: The `AsExpression` macro already generates impls for `&'a &'a` so that would not be the first case in diesel for such a construct.

> That works. Ideally we wanted to avoid deriving an additional impl for each derived struct but if that is an acceptable solution we will use this way in a possible PR.

Generally speaking: I'm fine with generating additional impls if this is required to workaround conflicting impl cases. If it would be possible to solve this case with a clever wild card impl that would be preferred. And again: If we want to add this impl we should add it now due to the fact that it would be a breaking change.

**Reply by valkum** [reply]
I opened a PR with the `Copy` bound blanket impl.
When using your version I hit a recursion limit in the BelongingToDsl blanket impl.
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9d12ee99d5bef9e78617bc35beb2465f )

Edit: Ah saw your PR right now. 
BTW: I am pretty sure the `Identifiable` blanket impl is alright, as we currently have struct in the tests that are Copy and rustc sees no overlapping impls for them. Do you see other problems with providing one for `Identifiable`?

**Reply by weiznich** [reply]
The blanket impl itself is fine, at least so that it does not cause overlapping impls. I had issues as soon as I've tried to use any of the method that depended on those blanket impl as rustc would go down a wrong route while inferring types. This results in error messages like that one:

```
error[E0275]: overflow evaluating the requirement `&&_: diesel::Identifiable`
  --> diesel_tests/tests/associations.rs:56:113
   |
56 |     let posts = <Post as BelongingToDsl<&[&User]>>::belonging_to(&users).load::<Post>(&mut connection).unwrap().grouped_by(&users);
   |                                                                                                                 ^^^^^^^^^^
   |
   = help: consider adding a `#![recursion_limit="2048"]` attribute to your crate (`integration_tests`)
   = note: required because of the requirements on the impl of `diesel::Identifiable` for `&&&_`
   = note: 1022 redundant requirements hidden
   = note: required because of the requirements on the impl of `diesel::Identifiable` for `&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&_`
   = note: required because of the requirements on the impl of `diesel::GroupedBy<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&_>` for `Vec<backend_specifics::Post>`
```

Those could be worked around by using UFCS, but that is not really an option, as that makes it much harder to write the corresponding code. That was for my proposed blanket impl, without copy.

For the impl with the `Copy` bound: That will cause conflicting impl errors if the user has a struct that is copy + manually implements `Identifiable` directly. Especially it will generate error messages that looks like diesel is broken, which is something I try to avoid if possible. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7963cddeccf74466c81586be79da527a) for more details.

**Reply by valkum** [reply]
Ah. Alright. I forgot that usecase. I will withdraw my PR then. Your PR looks fine.
